### PR TITLE
Better compatibility with sympy ImmutableMatrix (fixes #363)

### LINKED ISSUE #363: Inconsistent behaviour of ImmutableMatrix
The SymEngine tests in SymPy CI were not running properly. I'm reenabling them in https://github.com/sympy/sympy/pull/21992 but have found some possible regressions in SymEngine to do with the behaviour of ImmutableMatrix.

This is as expected:
```python
$ USE_SYMENGINE=1 python
Python 3.9.0 (v3.9.0:9cf6752276, Oct  5 2020, 11:29:23) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.core.backend import Matrix, ImmutableMatrix, sin, cos
>>> M = Matrix([[sin(1)**2 + cos(1)**2 - 1]])
>>> M
[-1 + sin(1)**2 + cos(1)**2]

>>> print(M.simplify())
None
>>> M
[0]
```
However this is not expected:
```python
>>> IM = ImmutableMatrix([[sin(1)**2 + cos(1)**2 - 1]])
>>> IM
[-1 + sin(1)**2 + cos(1)**2]

>>> print(IM.simplify())
None
>>> IM
[0]
```
It should not be possible to mutate the immutable matrix. SymPy's ImmutableMatrix class returns a simplified copy and does not mutate self:
```python
>>> from sympy import ImmutableMatrix
>>> SIM = ImmutableMatrix([[sin(1)**2 + cos(1)**2 - 1]])
>>> SIM
Matrix([[-1 + cos(1)**2 + sin(1)**2]])
>>> print(SIM.simplify())
Matrix([[0]])
>>> SIM
Matrix([[-1 + cos(1)**2 + sin(1)**2]])
```
It would have been better if the mutable matrix method `simplify` had a different name because it is inconsistent with almost all the other `.simplify()` methods in SymPy. For compatibility though the SymEngine wrapper should match this behaviour.

This would have shown up in SymPy CI if the SymEngine tests were still running properly (I noticed this from the old Travis CI).

Another inconsistency is in binary operations that mix and match mutable and immutable matrices. With SymPy we have:
```python
>>> from sympy import Matrix, ImmutableMatrix
>>> M1 = Matrix([[1]])
>>> M2 = ImmutableMatrix([[1]])
>>> for M in [M1, M2, M1+M2, M2+M1]: print(type(M))
... 
<class 'sympy.matrices.dense.MutableDenseMatrix'>
<class 'sympy.matrices.immutable.ImmutableDenseMatrix'>
<class 'sympy.matrices.immutable.ImmutableDenseMatrix'>
<class 'sympy.matrices.immutable.ImmutableDenseMatrix'>
```
In other words all operations mixing mutable and immutable coerce to immutable. With SymEngine that instead gives:
```python
>>> from sympy.core.backend import Matrix, ImmutableMatrix
>>> M1 = Matrix([[1]])
>>> M2 = ImmutableMatrix([[1]])
>>> for M in [M1, M2, M1+M2, M2+M1]: print(type(M))
... 
<class 'symengine.lib.symengine_wrapper.MutableDenseMatrix'>
<class 'symengine.lib.symengine_wrapper.ImmutableDenseMatrix'>
<class 'symengine.lib.symengine_wrapper.MutableDenseMatrix'>
<class 'symengine.lib.symengine_wrapper.ImmutableDenseMatrix'>
```
So the type of `M1+M2` does not match `M2+M1`.

### PR DESCRIPTION
**Pull request by rikardn**
* simplify Immutable to new Immutable
* sympy.sympify matrices to Immutable

### CODE REVIEW

File: symengine/lib/symengine_wrapper.pyx [hunk 1]
```diff
@@ -3897,7 +3897,7 @@ cdef class DenseMatrixBase(MatrixBase):
                 l.append(c2py(A.get(i, j))._sage_())
             s.append(l)
         import sage.all as sage
-        return sage.Matrix(s)
+        return sage.ImmutableMatrix(s)
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx:3900]:
```suggestion
        return sage.Matrix(s, immutable=True)
```

File: symengine/lib/symengine_wrapper.pyx [hunk 2]
```diff
@@ -4037,6 +4037,16 @@ cdef class ImmutableDenseMatrix(DenseMatrixBase):
...
         raise TypeError("Cannot set values of {}".format(self.__class__))
 
+    def _applyfunc(self, f):
+        cdef int nr = self.nrows()
+        cdef int nc = self.ncols()
+        temp = DenseMatrix(self)
+        for i in range(nr):
+            for j in range(nc):
+                temp._set(i, j, f(temp._get(i, j)))
+        return ImmutableDenseMatrix(temp)
```
Comment (isuruf) [inline symengine/lib/symengine_wrapper.pyx:4047]:
```suggestion
        res = DenseMatrix(self)
        res._applyfunc(f)
        return ImmutableDenseMatrix(res)
```

**Comment by isuruf** [comment]
Can you paste the output of the following with this PR?

```
import symengine
import sympy
print(type(symengine.zeros(5) + symengine.zeros(5)))
print(type(sympy.zeros(5) + sympy.zeros(5)))
print(type(symengine.zeros(5) + sympy.zeros(5)))
print(type(sympy.zeros(5) + symengine.zeros(5)))
```

**Comment by rikardn** [comment]
Sure:

```
n [1]: import symengine
   ...: import sympy
   ...: print(type(symengine.zeros(5) + symengine.zeros(5)))
   ...: print(type(sympy.zeros(5) + sympy.zeros(5)))
   ...: print(type(symengine.zeros(5) + sympy.zeros(5)))
   ...: print(type(sympy.zeros(5) + symengine.zeros(5)))
<class 'symengine.lib.symengine_wrapper.MutableDenseMatrix'>
<class 'sympy.matrices.dense.MutableDenseMatrix'>
<class 'symengine.lib.symengine_wrapper.MutableDenseMatrix'>
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-61b037069457> in <module>
      4 print(type(sympy.zeros(5) + sympy.zeros(5)))
      5 print(type(symengine.zeros(5) + sympy.zeros(5)))
----> 6 print(type(sympy.zeros(5) + symengine.zeros(5)))

~/testing/myvenv/lib/python3.8/site-packages/sympy/core/decorators.py in binary_op_wrapper(self, other)
    134                     if f is not None:
    135                         return f(self)
--> 136             return func(self, other)
    137         return binary_op_wrapper
    138     return priority_decorator

~/testing/myvenv/lib/python3.8/site-packages/sympy/matrices/common.py in __add__(self, other)
   2708             # call the highest-priority class's _eval_add
   2709             a, b = self, other
-> 2710             if a.__class__ != classof(a, b):
   2711                 b, a = a, b
   2712             return a._eval_add(b)

~/testing/myvenv/lib/python3.8/site-packages/sympy/matrices/common.py in classof(A, B)
   3283             return A.__class__
   3284 
-> 3285     raise TypeError("Incompatible classes %s, %s" % (A.__class__, B.__class__))

TypeError: Incompatible classes <class 'sympy.matrices.dense.MutableDenseMatrix'>, <class 'symengine.lib.symengine_wrapper.MutableDenseMatrix'>
```

**Comment by isuruf** [comment]
Merging without appveyor for now as the queue is long. Will revert if appveyor fails.

Thanks for the PR @rikardn

**Pull request merged**